File: crm/app/internal/adapters/inbound/http/campaign_handler.go
// Package http — Campaign HTTP handlers.
// Auth: DEFERRED — LAN-trust only (per plan).
package http

import (
	"context"
	"encoding/json"
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// CampaignReader covers read operations for campaigns.
type CampaignReader interface {
	GetCampaign(ctx context.Context, campaignID string) (*domain.Campaign, error)
	ListCampaigns(ctx context.Context) ([]domain.Campaign, error)
	ListTargets(ctx context.Context, campaignID, status string) ([]domain.CampaignTarget, error)
	GetROI(ctx context.Context, campaignID string) (*domain.CampaignROI, error)
}

// CampaignWriter covers campaign creation and mutation.
type CampaignWriter interface {
	CreateCampaign(ctx context.Context, c *domain.Campaign) error
	UpdateCampaign(ctx context.Context, campaignID string, patch *domain.Campaign) error
	GenerateTargets(ctx context.Context, campaignID string) (int, error)
	UpdateTargetStatus(ctx context.Context, campaignID, partyID, newStatus string) error
	AssignTarget(ctx context.Context, campaignID, partyID, userID string) error
	ScanConversions(ctx context.Context, campaignID string) (int, error)
}

// CampaignHandler wires campaign endpoints.
type CampaignHandler struct {
	reader CampaignReader
	writer CampaignWriter
}

// NewCampaignHandler constructs the handler.
func NewCampaignHandler(r CampaignReader, w CampaignWriter) *CampaignHandler {
	return &CampaignHandler{reader: r, writer: w}
}

// RegisterRoutes mounts campaign routes on the given chi router.
func (h *CampaignHandler) RegisterRoutes(r chi.Router) {
	r.Get("/api/campaigns", h.listCampaigns)
	r.Post("/api/campaigns", h.createCampaign)
	r.Patch("/api/campaigns/{id}", h.updateCampaign)
	r.Post("/api/campaigns/{id}/generate-targets", h.generateTargets)
	r.Get("/api/campaigns/{id}/targets", h.listTargets)
	r.Patch("/api/campaigns/{id}/targets/{partyId}", h.updateTarget)
	r.Post("/api/campaigns/{id}/scan-conversions", h.scanConversions)
	r.Get("/api/campaigns/{id}/roi", h.getROI)
}

// GET /api/campaigns
func (h *CampaignHandler) listCampaigns(w http.ResponseWriter, r *http.Request) {
	campaigns, err := h.reader.ListCampaigns(r.Context())
	if err != nil {
		log.Printf("campaign handler: list: %v", err)
		writeError(w, http.StatusInternalServerError, "failed to list campaigns")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"campaigns": campaigns})
}

// POST /api/campaigns
func (h *CampaignHandler) createCampaign(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name        string  `json:"name"`
		Objective   string  `json:"objective"`
		Channel     string  `json:"channel"`
		SegmentID   *string `json:"segment_id"`
		ScheduledAt *string `json:"scheduled_at"`
		CreatedBy   *string `json:"created_by"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	c := &domain.Campaign{
		Name:        body.Name,
		Objective:   body.Objective,
		Channel:     body.Channel,
		SegmentID:   body.SegmentID,
		ScheduledAt: body.ScheduledAt,
		CreatedBy:   body.CreatedBy,
	}
	if err := h.writer.CreateCampaign(r.Context(), c); err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("campaign handler: create: %v", err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusCreated, c)
}

// PATCH /api/campaigns/{id}
func (h *CampaignHandler) updateCampaign(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	var body struct {
		Name        string  `json:"name"`
		Objective   string  `json:"objective"`
		Channel     string  `json:"channel"`
		SegmentID   *string `json:"segment_id"`
		Status      string  `json:"status"`
		ScheduledAt *string `json:"scheduled_at"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	patch := &domain.Campaign{
		Name:        body.Name,
		Objective:   body.Objective,
		Channel:     body.Channel,
		SegmentID:   body.SegmentID,
		Status:      body.Status,
		ScheduledAt: body.ScheduledAt,
	}
	if err := h.writer.UpdateCampaign(r.Context(), campaignID, patch); err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("campaign handler: update %s: %v", campaignID, err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok", "campaign_id": campaignID})
}

// POST /api/campaigns/{id}/generate-targets
func (h *CampaignHandler) generateTargets(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	n, err := h.writer.GenerateTargets(r.Context(), campaignID)
	if err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("campaign handler: generate targets %s: %v", campaignID, err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"campaign_id": campaignID, "targets_created": n})
}

// GET /api/campaigns/{id}/targets?status=<status>
func (h *CampaignHandler) listTargets(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	status := r.URL.Query().Get("status")
	targets, err := h.reader.ListTargets(r.Context(), campaignID, status)
	if err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("campaign handler: list targets %s: %v", campaignID, err)
			writeError(w, http.StatusInternalServerError, "failed to list targets")
		}
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"targets": targets})
}

// PATCH /api/campaigns/{id}/targets/{partyId}
// body: {"status":"sent"} or {"assigned_user_id":"..."}
func (h *CampaignHandler) updateTarget(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	partyID := chi.URLParam(r, "partyId")

	var body struct {
		Status         *string `json:"status"`
		AssignedUserID *string `json:"assigned_user_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if body.Status != nil {
		if err := h.writer.UpdateTargetStatus(r.Context(), campaignID, partyID, *body.Status); err != nil {
			if domain.IsValidationError(err) {
				writeError(w, http.StatusBadRequest, err.Error())
			} else {
				log.Printf("campaign handler: update target status (%s,%s): %v", campaignID, partyID, err)
				writeError(w, http.StatusInternalServerError, "operation failed")
			}
			return
		}
	}
	if body.AssignedUserID != nil {
		if err := h.writer.AssignTarget(r.Context(), campaignID, partyID, *body.AssignedUserID); err != nil {
			if domain.IsValidationError(err) {
				writeError(w, http.StatusBadRequest, err.Error())
			} else {
				log.Printf("campaign handler: assign target (%s,%s): %v", campaignID, partyID, err)
				writeError(w, http.StatusInternalServerError, "operation failed")
			}
			return
		}
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// POST /api/campaigns/{id}/scan-conversions
func (h *CampaignHandler) scanConversions(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	n, err := h.writer.ScanConversions(r.Context(), campaignID)
	if err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("campaign handler: scan conversions %s: %v", campaignID, err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"campaign_id": campaignID, "newly_converted": n})
}

// GET /api/campaigns/{id}/roi
func (h *CampaignHandler) getROI(w http.ResponseWriter, r *http.Request) {
	campaignID := chi.URLParam(r, "id")
	roi, err := h.reader.GetROI(r.Context(), campaignID)
	if err != nil {
		log.Printf("campaign handler: roi %s: %v", campaignID, err)
		writeError(w, http.StatusInternalServerError, "failed to compute ROI")
		return
	}
	writeJSON(w, http.StatusOK, roi)
}

File: crm/src/adapters/inbound/http/campaign_handler.py
"""Campaign HTTP handlers — FastAPI APIRouter prefix=/api.

Auth: DEFERRED — LAN-trust only (per plan).
"""
from __future__ import annotations

import logging
from dataclasses import asdict
from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class CampaignCreateBody(BaseModel):
    name: str
    objective: str
    channel: str
    segment_id: Optional[str] = None
    scheduled_at: Optional[str] = None
    created_by: Optional[str] = None


class CampaignUpdateBody(BaseModel):
    name: Optional[str] = None
    objective: Optional[str] = None
    channel: Optional[str] = None
    segment_id: Optional[str] = None
    status: Optional[str] = None
    scheduled_at: Optional[str] = None


class TargetUpdateBody(BaseModel):
    status: Optional[str] = None
    assigned_user_id: Optional[str] = None


# ---------------------------------------------------------------------------
# Router factory
# ---------------------------------------------------------------------------

def make_campaign_router(campaign_svc: Any) -> APIRouter:
    """Return an APIRouter wired to *campaign_svc* (CampaignService instance).

    Keeps the handler thin: validation and business rules live in the service layer.
    """
    router = APIRouter(prefix="/api")

    # ── GET /campaigns ───────────────────────────────────────────────────────

    @router.get("/campaigns")
    def list_campaigns():
        try:
            campaigns = campaign_svc.list_campaigns()
        except Exception as exc:
            logger.exception("campaign handler: list")
            raise HTTPException(status_code=500, detail="failed to list campaigns") from exc
        return {"campaigns": [asdict(c) for c in campaigns]}

    # ── POST /campaigns ──────────────────────────────────────────────────────

    @router.post("/campaigns", status_code=201)
    def create_campaign(body: CampaignCreateBody):
        try:
            campaign = campaign_svc.create_campaign(body.model_dump())
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        except Exception as exc:
            logger.exception("campaign handler: create")
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return asdict(campaign)

    # ── PATCH /campaigns/{id} ────────────────────────────────────────────────

    @router.patch("/campaigns/{campaign_id}")
    def update_campaign(campaign_id: str, body: CampaignUpdateBody):
        patch = {k: v for k, v in body.model_dump().items() if v is not None}
        try:
            campaign_svc.update_campaign(campaign_id, **patch)
        except ValueError as exc:
            msg = str(exc)
            status = 404 if "not found" in msg else 400
            raise HTTPException(status_code=status, detail=msg) from exc
        except Exception as exc:
            logger.exception("campaign handler: update %s", campaign_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"status": "ok", "campaign_id": campaign_id}

    # ── POST /campaigns/{id}/generate-targets ────────────────────────────────

    @router.post("/campaigns/{campaign_id}/generate-targets")
    def generate_targets(campaign_id: str):
        try:
            n = campaign_svc.generate_targets(campaign_id)
        except ValueError as exc:
            msg = str(exc)
            status = 404 if "not found" in msg else 400
            raise HTTPException(status_code=status, detail=msg) from exc
        except Exception as exc:
            logger.exception("campaign handler: generate targets %s", campaign_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"campaign_id": campaign_id, "targets_created": n}

    # ── GET /campaigns/{id}/targets ──────────────────────────────────────────

    @router.get("/campaigns/{campaign_id}/targets")
    def list_targets(campaign_id: str, status: Optional[str] = Query(default=None)):
        try:
            targets = campaign_svc.list_targets(campaign_id, status)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        except Exception as exc:
            logger.exception("campaign handler: list targets %s", campaign_id)
            raise HTTPException(status_code=500, detail="failed to list targets") from exc
        return {"targets": [asdict(t) for t in targets]}

    # ── PATCH /campaigns/{id}/targets/{party_id} ─────────────────────────────
    # body: {"status": "sent"} or {"assigned_user_id": "..."} or both

    @router.patch("/campaigns/{campaign_id}/targets/{party_id}")
    def update_target(campaign_id: str, party_id: str, body: TargetUpdateBody):
        patch = {k: v for k, v in body.model_dump().items() if v is not None}
        if not patch:
            raise HTTPException(status_code=400, detail="no fields to update")
        try:
            campaign_svc.update_target(campaign_id, party_id, **patch)
        except ValueError as exc:
            msg = str(exc)
            status_code = 404 if "not found" in msg else 400
            raise HTTPException(status_code=status_code, detail=msg) from exc
        except Exception as exc:
            logger.exception(
                "campaign handler: update target (%s, %s)", campaign_id, party_id
            )
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"status": "ok"}

    # ── POST /campaigns/{id}/scan-conversions ────────────────────────────────

    @router.post("/campaigns/{campaign_id}/scan-conversions")
    def scan_conversions(campaign_id: str):
        try:
            n = campaign_svc.scan_conversions(campaign_id)
        except ValueError as exc:
            msg = str(exc)
            status = 404 if "not found" in msg else 400
            raise HTTPException(status_code=status, detail=msg) from exc
        except Exception as exc:
            logger.exception("campaign handler: scan conversions %s", campaign_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"campaign_id": campaign_id, "newly_converted": n}

    # ── GET /campaigns/{id}/roi ──────────────────────────────────────────────

    @router.get("/campaigns/{campaign_id}/roi")
    def get_roi(campaign_id: str):
        try:
            roi = campaign_svc.get_roi(campaign_id)
        except Exception as exc:
            logger.exception("campaign handler: roi %s", campaign_id)
            raise HTTPException(status_code=500, detail="failed to compute ROI") from exc
        return roi

    return router
